### DISCUSSION TITLE
EntityRuler with LEMMA pattern results in errors [E155], [E1004], [E050]

### DISCUSSION THREAD

**Discussion by afparsons** [explosion/spaCy#7430] [Help: Coding & Implementations] [answered]
I have read through the [usage guide for EntityRuler](https://spacy.io/usage/rule-based-matching#entityruler).
> Entity patterns are dictionaries with two keys: "label" . . . and "pattern", the match pattern.
> ...
> Token patterns with one dictionary describing one token (list).

Awesome! These appear to be the same types of patterns idenitied in the [usage guide section, "Available Token Attributes"](https://spacy.io/usage/rule-based-matching#adding-patterns-attributes).

Therefore, I understand that I can use `LEMMA` in an EntityRuler pattern, like so:

```python
patterns: List[Dict] = [
    {
        'label': '<SOME_LABEL>',
        'id': '<SOME_ID>',
        'pattern': [{'LEMMA': {'IN': ['like', 'love']}}, {'POS': 'NOUN'}]
    }
]

entity_ruler = nlp.add_pipe('entity_ruler')
entity_ruler.add_patterns(patterns)
```

However, using the `LEMMA` token attribute causes errors. This probably comes down to a simple misunderstanding, but nonetheless, I am confused.

In the following section, I include a large code block, followed by successive excerpts from my attempts to fix the issue and the resulting errors.

---

```python
import json
import spacy

# the library I am developing
from haleynlp.extraction.en.chronometry import EN_DURATIONS

nlp = spacy.load('en_core_web_md')

# patterns stored in a JSON file; yes, I'm aware of JSONL, but manually editing such files proves cumbersome
with open('../../haleynlp/patterns/dates.json', 'r') as f:
    patterns = json.load(f)

# where key is in {'week', 'weekly', 'biweekly', 'daily', 'day', 'hour', 'hourly', etc.}
patterns.extend([
    {
        'id': key,
        'label': 'CHRONO',
        'pattern': [
            {'LOWER': {'IN': ['a', 'by the', 'every', 'per']}},
            {'LEMMA': key}  # should be LEMMA
        ][0 if not key.endswith('ly') else 1:]
    }
    for key in EN_DURATIONS.keys()
])

component_entity_ruler = nlp.add_pipe(factory_name='entity_ruler', before='ner')
component_entity_ruler.add_patterns(patterns)
```

### 1. "Oh, the Lemmatizer needs to come before the EntityRuler"

```python
>>> doc = nlp('You will be charged hourly.')

ValueError: [E155] The pipeline needs to include a lemmatizer in order to use Matcher or PhraseMatcher with the attribute LEMMA. Try using `nlp()` instead of `nlp.make_doc()` or `list(nlp.pipe())` instead of `list(nlp.tokenizer.pipe())`.

>>> nlp.pipe_names

['tok2vec',
 'tagger',
 'parser',
 'lemmatizer',
 'entity_ruler',
 'ner',
 'attribute_ruler']
```

### 2. "Hmm, ok"

Seeing the above, I thought that moving the Lemmatizer to come before the EntityRuler might help:

```python
...
nlp = spacy.load('en_core_web_md', exclude='lemmatizer')
...
nlp.add_pipe('lemmatizer', before='ner')
...

>>> nlp.pipe_names

['tok2vec',
 'tagger',
 'parser',
 'lemmatizer',
 'entity_ruler',
 'ner',
 'attribute_ruler']

>>> doc = nlp('You will be charged hourly.')

ValueError: [E1004] Missing lemmatizer table(s) found for lemmatizer mode 'rule'. Required tables: ['lemma_rules']. Found: []. Maybe you forgot to call `nlp.initialize()` to load in the data?
```

### 3. "I'll try this, but I don't understand why"

```python
...
# at the end of the code block
nlp.initialize()

OSError: [E050] Can't find model 'corpus/en_vectors'. It doesn't seem to be a Python package or a valid path to a data directory.
```

What is going on here, and what am I doing wrong?
For what it is worth, replacing `LEMMA` with `LOWER` works just fine; however, I need `"hours" --> "hour"`.

---

I know the team at Explosion has already been showered with praise, but for what it is worth, I'd also like to add my acclimation: spaCy has been a pleasure to use! Thanks for all of your hard work.

---

Edit 1: spelling

**Comment by polm** [comment]
Huh, that is weird. It looks like there are a couple of weird things happening:

1. The new Lemmatizer you make is trying to be rule-based
2. The new Lemmatizer needs to be initialized to load its data

2 is not obvious but it is expected behavior. 1 does not match the docs or source, not sure if I'm misunderstanding something or if that's an issue, I'll need to look into that more.

I think this code is close to what you were trying to do:

```
import spacy

nlp = spacy.load("en_core_web_sm", exclude="lemmatizer")

DURATIONS = 'week weekly biweekly daily day'.split()

patterns = [
    {
        'id': key,
        'label': 'CHRONO',
        'pattern': [
            {'LOWER': {'IN': ['a', 'by the', 'every', 'per']}},
            {'LEMMA': key}  # should be LEMMA
        ][0 if not key.endswith('ly') else 1:]
    }
    for key in DURATIONS
]

texts = [
        "I water the plants daily.",
        "When is the weekly meeting?"
]

lemma_config = {"mode":"lookup"}
lemmatizer = nlp.add_pipe("lemmatizer", before="ner", config=lemma_config)
lemmatizer.initialize()
ruler = nlp.add_pipe("entity_ruler", before="ner")
ruler.add_patterns(patterns)

print(patterns)
for text in texts:
    for ent in nlp(text).ents:
        print(ent, ent.label_)
```

One comment here is that rather than matching on the `a, by the, every` part you may be able to match more easily on the part of speech of the key word.

**Reply by polm** [reply]
Ah, 1 is the case because the English language overrides the Lemmatizer defaults to be rule based instead of lookup based. So that explains that.
 
https://github.com/explosion/spaCy/blob/be44257cab314cadf66cdd4805fccdaa36039522/spacy/lang/en/__init__.py#L29

**Comment by adrianeboyd** [comment, marked as answer]
Sorry, let me jump in here -- this is the same issue as #7382. The easiest method is to move `ner` to the end of the pipe, after `attribute_ruler` and `lemmatizer`. The rule-based lemmatizer depends on `tagger` + `attribute_ruler` to work correctly. You can use the `source` option to load the same component (same config + same weights/rules/data) in a new position in the pipeline.

```python
import spacy
nlp = spacy.load("en_core_web_sm")

# move the NER component to the end of the pipeline: remove and then reload from the same source in the new position
nlp.remove_pipe("ner")
nlp.add_pipe("ner", source=spacy.load("en_core_web_sm"))

# add entity ruler
nlp.add_pipe("entity_ruler", before="ner")
```

Moving pipeline components needs to be easier (we could add something like `nlp.move_pipe`?) and I think we should move `ner` to the end in the v3.1 pretrained pipelines because this is a headache for users. `attribute_ruler`+`lemmatizer` use annotation from `tagger` and `parser`, but not `ner`.

**Comment by afparsons** [comment]
Thank you to both @adrianeboyd and @polm! You both were very quick to answer.
